commit d8b920afe7765621bf122f63a585ece9b3a64532 (refs/changes/25/1725/5)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-04-10T14:57:07+00:00 (2 years, 6 months ago)
    
    TOOLS-1731 sdcadm post-setup common-external-nics should set external network as primary
    Reviewed by: Trent Mick <trentm@gmail.com>

diff --git a/lib/cli/do_update_other.js b/lib/cli/do_update_other.js
index f019750..5a471ce 100644
--- a/lib/cli/do_update_other.js
+++ b/lib/cli/do_update_other.js
@@ -755,6 +755,79 @@ function do_update_other(subcmd, opts, args, cb) {
             }, next);
         },
 
+        /*
+         * Clean up after older vers of 'sdcadm post-setup common-external-nics'
+         * that did not set the "external" NIC to be the primary.
+         */
+        function updateCommonExternalNics(ctx, next) {
+            var initialNetworks = ['admin'];
+            var oldNetworks = ['admin', 'external'];
+            var newNetworks = [
+                { name: 'admin' },
+                { name: 'external', primary: true}
+            ];
+            var toUpdateItems = [];
+
+            var svcs = ctx.svcs.filter(function (svc) {
+                return (svc.name === 'adminui' || svc.name === 'imgapi');
+            });
+
+            self.sdcadm.checkMissingExternalNics({
+                progress: progress
+            }, function (sdcadmErr, vmsNics) {
+                if (sdcadmErr) {
+                    next(sdcadmErr);
+                    return;
+                }
+
+                /*
+                 * vmsNics has two members: doimgapi and doadminui which
+                 * will be true if adding the nics to imgapi0 or adminui0 is
+                 * still pending and, on that case, we can safely skip the
+                 * service here, otherwise, we need to update params.
+                 */
+                svcs.forEach(function (svc) {
+                    if (!vmsNics['do' + svc.name] &&
+                        svc.params &&
+                        (jsprim.deepEqual(svc.params.networks, oldNetworks) ||
+                        jsprim.deepEqual(svc.params.networks, initialNetworks))
+                    ) {
+                        toUpdateItems.push({
+                            svcName: svc.name,
+                            svcUuid: svc.uuid,
+                            paramsNetworks: newNetworks
+                        });
+                    }
+
+                    if (toUpdateItems.length === 0) {
+                        next();
+                        return;
+                    }
+                });
+
+                vasync.forEachPipeline({
+                    inputs: toUpdateItems,
+                    func: function handleOneUpdateItem(item, nextItem) {
+                        progress('Update "%s" service params.networks to ' +
+                            'ensure the external NIC is primary',
+                            item.svcName);
+                        updateService(item.svcUuid, {
+                            params: {
+                                networks: item.paramsNetworks
+                            }
+                        }, function (updateErr) {
+                            if (updateErr) {
+                                nextItem(new errors.SDCClientError(
+                                    updateErr, 'sapi'));
+                            } else {
+                                nextItem();
+                            }
+                        });
+                    }
+                }, next);
+            });
+        },
+
         steps.agentServicesEnsureCreated
 
     ]}, function (err) {
diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 3b99fbd..6577211 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -6212,7 +6212,10 @@ setupCommonExternalNics(opts, cb) {
 
             self.sapi.updateService(svcArr[0].uuid, {
                 params: {
-                    networks: ['admin', 'external']
+                    networks: [
+                        { name: 'admin' },
+                        { name: 'external', primary: true}
+                    ]
                 }
             }, function (err) {
                 if (err) {
